### PR TITLE
Add integration with pgsql in vagga

### PR DESCRIPTION
**Pull request by DrEdi**
Add new app where we create admin for db
Integrate vagga with pgsql
Refresh requirements

### CODE REVIEW

**Review by Miseracle — APPROVED** [review]

File: common/migrations/0001_create_superuser.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+"""Module for creating DB-admin."""
+from django.db import migrations
+from django.contrib.auth.hashers import make_password
+
+
+def create_superuser(apps, schema_editor):
+    """Module is used when we start container with PgSQL."""
```
Comment (Miseracle) [inline common/migrations/0001_create_superuser.py]: Is it a docstring for a module?

**Review by Miseracle — APPROVED** [review]

File: common/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""Empty file."""
```
Comment (webknjaz) [inline common/__init__.py]: Why do you need this file?

File: common/migrations/0001_create_superuser.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    User = apps.get_model('auth', 'User')
+    User.objects.create(username='admin',
+                        email='admin@example.com',
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: I'd read this from environment with a default fallback. Like `os.getenv('ADMIN_EMAIL', 'admin@example.com')`

Comment (webknjaz) [reply common/migrations/0001_create_superuser.py]: (or define it via `django-environ` in settings)

File: common/migrations/0001_create_superuser.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+                        password=make_password('12345678'),
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py:12]: I would generate random password

File: common/migrations/0001_create_superuser.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+                        is_superuser=True,
+                        is_staff=True,
+                        is_active=True)
+
+
+class Migration(migrations.Migration):
+    """Config class."""
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: I think it's a migration plan for superuser creation.

File: common/__init__.py (hunk 2 above)
Comment (webknjaz) [reply common/__init__.py]: Current docstring does not make sense, but you can make it useful by adding a docstring describing the whole `common` package

File: common/migrations/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+"""Empty file."""
```
Comment (webknjaz) [inline common/migrations/__init__.py]: Why did you create this file?

File: vagga.yaml [hunk 7]
```diff
@@ -2,19 +2,41 @@ containers:
   pip-freeze:
     setup:
       - !Alpine v3.4
+      - !BuildDeps
+        - postgresql-dev
```
Comment (webknjaz) [inline vagga.yaml:7]: Why it's not `libpq` as in `django` container?

File: vagga.yaml [hunk 8]
```diff
@@ -2,19 +2,41 @@ containers:
...
     setup:
       - !Alpine v3.4
       - !PipConfig
         dependencies: true
+      - !Install
+        - libpq
       - !Py3Requirements requirements.txt
+  postgres:
+    setup:
+    - !Ubuntu trusty
```
Comment (webknjaz) [inline vagga.yaml]: JFYI there's newer Ubuntu version, which is `xenial`

File: vagga.yaml [hunk 9]
```diff
@@ -24,3 +46,26 @@ commands:
     description: Shortcut to manage.py
     container: django
     run: [python3, manage.py]
+  run-postgres: !Supervise
+    description: Start the django development server using Postgres database
+    children:
+      app: !Command
+        container: django
+        environ:
```
Comment (webknjaz) [inline vagga.yaml]: Nit: you could add this right into container description, since this is not the only command needing this var

File: vagga.yaml [hunk 10]
```diff
@@ -24,3 +46,26 @@ commands:
...
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/django
+        run: |
+            touch /work/.dbcreation # Create lock file
```
Comment (webknjaz) [inline vagga.yaml]: I would follow python's conventions here as well and use 2 whitespaces before `#`

File: vagga.yaml (hunk 8 above)
Comment (webknjaz) [reply vagga.yaml]: P.S. You could also base it on Alpine

File: vagga.yaml [hunk 11]
```diff
@@ -24,3 +46,26 @@ commands:
...
+            while [ -f /work/.dbcreation ]; do sleep 0.2; done # Acquire lock
+            python3 manage.py migrate
+            python3 manage.py runserver --insecure
+      db: !Command
+        container: postgres
+        run: |
+            chown postgres:postgres $PGDATA;
+            su postgres -c "$PG_BIN/pg_ctl initdb";
```
Comment (webknjaz) [inline vagga.yaml]: I think, DB creation should not repeat on every command run, you need a separate command for initializing the db

File: vagga.yaml [hunk 12]
```diff
@@ -2,19 +2,41 @@ containers:
...
+    - !EnsureDir /data
+    environ:
+      PGDATA: /data
+      PG_PORT: 5433
+      PG_DB: django
+      PG_USER: vagga
+      PG_PASSWORD: vagga
+      PG_BIN: /usr/lib/postgresql/9.3/bin
+    volumes:
+      /data: !Tmpfs
```
Comment (webknjaz) [inline vagga.yaml]: Use persistent named volume

File: vagga.yaml (hunk 11 above)
Comment (webknjaz) [reply vagga.yaml]: Look for init command in volume definition

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: common/migrations/__init__.py (hunk 6 above)
Comment (DrEdi) [reply common/migrations/__init__.py]: Auto created
Would it better to remove?

Comment (webknjaz) [reply common/migrations/__init__.py]: As u wish, but remove that stupid docstring

File: common/migrations/0001_create_superuser.py (hunk 4 above)
Comment (DrEdi) [reply common/migrations/0001_create_superuser.py:12]: But how will I know this random generated password?
From environment?

File: vagga.yaml (hunk 12 above)
Comment (DrEdi) [inline vagga.yaml]: Opps
`vagga.yaml:24:5: Validation Error: Keys {"volume"} are not expected`

File: vagga.yaml (hunk 7 above)
Comment (DrEdi) [reply vagga.yaml:7]: We need it in runtime

Comment (webknjaz) [reply vagga.yaml:7]: so why it's different here?

Comment (DrEdi) [reply vagga.yaml:7]: Documentations says that postgresql needed for building psycopg, libpq - needed for it in runtime.
There's no other info.
But I think difference is also in !BuildDeps and !Install

File: common/migrations/0001_create_superuser.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+"""Module for creating DB-admin."""
+from django.db import migrations
+from django.contrib.auth.hashers import make_password
+import environ
+
+env = environ.Env(DEBUG=(bool, False),)
+
+
+def create_superuser(apps, schema_editor):
+    """Function is used when we start container with PgSQL."""
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: Write in imperative style:
```python
"""Creates superuser for the app at the first start"""
```

File: common/migrations/0001_create_superuser.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
+"""Module for creating DB-admin."""
+from django.db import migrations
+from django.contrib.auth.hashers import make_password
+import environ
+
+env = environ.Env(DEBUG=(bool, False),)
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: 1) Please don't forget reading `.env` into environment.
2) `DEBUG` var is not needed here.
3) It is better to move declaration of `ADMIN_EMAIL` into settings and just import it from there. Keeping all related things in one place helps a lot :)

File: vagga.yaml [hunk 15]
```diff
@@ -2,20 +2,65 @@ containers:
...
       - !Alpine v3.4
       - !PipConfig
         dependencies: true
+      - !Install
+        - libpq
       - !Py3Requirements requirements.txt
+  postgres:
+    setup:
+    - !Ubuntu xenial
+    - !EnsureDir /data
```
Comment (webknjaz) [inline vagga.yaml:31]: According to doc for `!Persistent`, `!EnsureDir` is not needed for such mount point. (Please check this)

File: vagga.yaml (hunk 7 above)
Comment (webknjaz) [reply vagga.yaml:7]: `!BuildDeps` are removed after container is built. `!Install` packages are available in runtime.
But the question here is that you install `psycopg2` python package and it requires some package during build time which you don't provide in `django` container. This means that `libpq` is enough for build, so why not use the same package here?

File: vagga.yaml [hunk 16]
```diff
@@ -24,3 +69,18 @@ commands:
...
+        environ:
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/test
+        run: |
+            python3 manage.py migrate
+            python3 manage.py runserver
+      db: !Command
+        container: postgres
+        user-id: 200
+        group-id: 200
+        run: exec $PG_BIN/postgres -F --port=$PG_PORT
```
Comment (webknjaz) [inline vagga.yaml]: to use EXEC syntax in vagga you may just pass an array. like:
```yaml
run: [exec, $PG_BIN/postgres, -F, --port=$PG_PORT]
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
I've pointed out some suggestions, but I'll need to return and check the logic of PG commands

File: vagga.yaml (hunk 15 above)
Comment (DrEdi) [reply vagga.yaml:31]: I have an error without this 
`No such file or directory (os error 2) (source: exists, target: missing, mapped-root)`

File: vagga.yaml (hunk 16 above)
Comment (DrEdi) [reply vagga.yaml]: This also come with error 
`ERROR:vagga::wrapper: Error executing run-postgres: Command exec not found in "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
---------- Process db:8849 exited with code 124. Shutting down -----------`

Comment (webknjaz) [reply vagga.yaml]: my bad, remove `[exec]` from the sequence

File: common/migrations/0001_create_superuser.py [hunk 17]
```diff
@@ -3,14 +3,15 @@
...
+environ.Env.read_env()
 
 
 def create_superuser(apps, schema_editor):
-    """Function is used when we start container with PgSQL."""
+    """Create superuser for the app at the first start."""
     User = apps.get_model('auth', 'User')
     User.objects.create(username='admin',
-                        email=env('ADMIN_EMAIL', 'admin@example.com'),
+                        email=env('ADMIN_EMAIL'),
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: why did you remove default value?

File: common/migrations/0001_create_superuser.py [hunk 18]
```diff
@@ -3,14 +3,15 @@
 from django.contrib.auth.hashers import make_password
 import environ
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: don't use `environ` in this module

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: vagga.yaml (hunk 16 above)
Comment (DrEdi) [reply vagga.yaml]: Something gone wrong
`ERROR:vagga::wrapper: Error executing run-postgres: Error running <Command "$PG_BIN/postgres" "-F" "--port=$PG_PORT";`

Comment (webknjaz) [reply vagga.yaml]: It looks like env vars are not getting interpolated in exec, so leave it as is for now.

File: common/migrations/0001_create_superuser.py [hunk 19]
```diff
@@ -0,0 +1,27 @@
+"""Module for creating DB-admin."""
+from django.db import migrations
+from django.contrib.auth.hashers import make_password
+from blog.settings import ADMIN_EMAIL
```
Comment (webknjaz) [inline common/migrations/0001_create_superuser.py]: https://docs.djangoproject.com/en/1.10/topics/settings/#using-settings-in-python-code

File: vagga.yaml [hunk 20]
```diff
@@ -2,20 +2,65 @@ containers:
...
+    - !Sh |
+        addgroup --system --gid 200 postgres
+        adduser --uid 200 --system --home /data --no-create-home \
+            --shell /bin/bash --group --gecos "PostgreSQL administrator" \
+            postgres
+    - !Install [postgresql-9.5]
+    environ:
+      PGDATA: /data
+      PG_PORT: 5433
+      PG_DB: test
```
Comment (webknjaz) [inline vagga.yaml]: Don't use test DB, create your own.

File: vagga.yaml [hunk 21]
```diff
@@ -24,3 +69,18 @@ commands:
...
     run: [python3, manage.py]
+  run-postgres: !Supervise
+    description: Start the django development server using Postgres database
+    children:
+      app: !Command
+        container: django
+        environ:
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/test
+        run: |
+            python3 manage.py migrate
```
Comment (webknjaz) [inline vagga.yaml:72]: You should normally wait until PG becomes available to be able to apply migrations.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: common/migrations/__init__.py [hunk 22]
```diff
@@ -0,0 +1 @@
+"""This file is needed for migration admin to DB."""
```
Comment (webknjaz) [inline common/migrations/__init__.py]: Describe the migrations package instead.

File: vagga.yaml [hunk 23]
```diff
@@ -75,7 +75,7 @@ commands:
       app: !Command
         container: django
         environ:
-          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/test
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
```
Comment (webknjaz) [inline vagga.yaml]: Why didn't you move it to container declaration?

File: blog/settings.py [hunk 24]
```diff
@@ -111,17 +110,8 @@
 # Static files (CSS, JavaScript, Images)
 # https://docs.djangoproject.com/en/1.10/howto/static-files/
 
-public_root = root.path('public/')
-
+public_root = env.path('STATIC_ROOT', BASE_DIR('static'))
```
Comment (webknjaz) [inline blog/settings.py]: I'm not sure whether this is correct

**Review by webknjaz — CHANGES_REQUESTED** [review]
Please read my old comments, which are collapsed by now as well

File: blog/settings.py (hunk 24 above)
Comment (DrEdi) [reply blog/settings.py]: It works. I've tested it before push

File: vagga.yaml (hunk 23 above)
Comment (DrEdi) [reply vagga.yaml]: It will not work if I move it to container:
`Is the server running on host "127.0.0.1" and accepting
	TCP/IP connections on port 8458?
`

File: vagga.yaml [hunk 25]
```diff
@@ -63,24 +63,22 @@ commands:
...
-    description: Start the django development server using Postgres database
-    children:
-      app: !Command
+  run-app: &run-app !Command
         container: django
         run: |
             python3 manage.py migrate
             python3 manage.py runserver --insecure
-      db: !Command
+  db: &run-db !Command
```
Comment (webknjaz) [inline vagga.yaml]: run-db plz

File: common/migrations/__init__.py [hunk 26]
```diff
@@ -0,0 +1,7 @@
+"""
```
Comment (webknjaz) [inline common/migrations/__init__.py:1]: Mind your grammar.

File: blog/settings.py (hunk 24 above)
Comment (webknjaz) [reply blog/settings.py]: So what does this invocation mean?

File: blog/settings.py [hunk 27]
```diff
@@ -3,11 +3,10 @@
...
 
 env = environ.Env(DEBUG=(bool, False),)
-root = environ.Path(__file__) - 3
-environ.Env.read_env()
-# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
-BASE_DIR = root()
 
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = environ.Path(__file__) - 2
+env.read_env(str(BASE_DIR.path('.env')))
```
Comment (webknjaz) [inline blog/settings.py]: Have you tried doing it [this way](https://github.com/open-craft-guild/crm-smsfly/blob/master/SMSFlyCRM/settings.py#L17)?

Comment (DrEdi) [reply blog/settings.py]: Yes. I have the same problem as before
`UserWarning: /work/blog/.env doesn't exist - if you're not configuring your environment separately, create one.
  "environment separately, create one." % env_file)`

File: vagga.yaml [hunk 28]
```diff
@@ -2,25 +2,83 @@ containers:
...
+  _pg-init: !Command
+    description: Init postgres database
+    container: postgres
+    user-id: 200
+    group-id: 200
+    run: |
+      set -ex
+      ls -la /data
+      $PG_BIN/pg_ctl initdb
+      $PG_BIN/pg_ctl -w -o '-F --port=$PG_PORT -k /tmp' start
```
Comment (webknjaz) [inline vagga.yaml]: I `/tmp` should be `/data`. Does it preserve DB across restarts?

File: vagga.yaml [hunk 29]
```diff
@@ -2,25 +2,83 @@ containers:
...
+      PG_USER: vagga
+      PG_PASSWORD: vagga
+      PG_BIN: /usr/lib/postgresql/9.5/bin
+    volumes:
+      /data: !Persistent
+        name: postgres
+        owner-uid: 200
+        owner-gid: 200
+        init-command: _pg-init
+      /run: !Tmpfs
```
Comment (webknjaz) [inline vagga.yaml:51]: Why did you add this?

File: vagga.yaml [hunk 30]
```diff
@@ -2,25 +2,83 @@ containers:
...
+  run-app: &run-app !Command
+        container: django
+        run: |
+            python3 manage.py migrate
+            python3 manage.py runserver --insecure
+  run-db: &run-db !Command
+        container: postgres
+        user-id: 200
+        group-id: 200
+        run: exec $PG_BIN/postgres -F --port=$PG_PORT
```
Comment (webknjaz) [inline vagga.yaml]: Why don't you pass `-k ...` to this command?

File: vagga.yaml (hunk 28 above)
Comment (DrEdi) [reply vagga.yaml]: Yes, everything is working ok

File: vagga.yaml (hunk 29 above)
Comment (DrEdi) [reply vagga.yaml:51]: It is in vagga documentation. I thought it'd be better not to change anything

Comment (webknjaz) [reply vagga.yaml:51]: It seems to be working fine w/o it

File: vagga.yaml (hunk 28 above)
Comment (webknjaz) [reply vagga.yaml]: Please clean volumes via `vagga _clean --volumes` and check again.

Comment (DrEdi) [reply vagga.yaml]: I've got fully clean project. But after add and reboot - new post still there, it's ok?

Comment (webknjaz) [reply vagga.yaml]: yes

File: blog/settings.py (hunk 27 above)
Comment (webknjaz) [reply blog/settings.py]: Have u tried leaving `environ.Env.read_env()`?

Comment (DrEdi) [reply blog/settings.py]: Yep. I always have a warning like in previous comment.

Comment (webknjaz) [reply blog/settings.py]: why ` - 2`?

Comment (DrEdi) [reply blog/settings.py]: I tested it ' by trial and error'. This is the only one thing where it sees .env file.
Also in stackoverfow where I found how to fix it was this variant.

File: vagga.yaml [hunk 31]
```diff
@@ -2,25 +2,83 @@ containers:
...
       - !Sh pip freeze > requirements.txt
   django:
     setup:
       - !Alpine v3.4
       - !PipConfig
         dependencies: true
+      - !Install
+        - libpq
       - !Py3Requirements requirements.txt
+    environ:
```
Comment (webknjaz) [inline vagga.yaml:25]: I would also add `environment-file` directive to this container declaration.

File: common/migrations/__init__.py [hunk 32]
```diff
@@ -0,0 +1,7 @@
...
+Package for creating admin account.
```
Comment (webknjaz) [inline common/migrations/__init__.py]: Fixture for admin account.

File: common/migrations/__init__.py [hunk 33]
```diff
@@ -0,0 +1,7 @@
...
+
+When you run project with pgsqlDB, admin model
```
Comment (webknjaz) [inline common/migrations/__init__.py]:
```
This will automatically add pre-configured admin account
to your db during migration. Specify admin's email via
ADMIN_EMAIL environment variable. The password is
hardcoded to 12345678, don't forget to reset it.
```

Comment (DrEdi) [reply common/migrations/__init__.py]: Changed.
Thanks

File: vagga.yaml (hunk 25 above)
Comment (DrEdi) [reply vagga.yaml]: done

File: common/migrations/__init__.py (hunk 22 above)
Comment (DrEdi) [reply common/migrations/__init__.py]: done

File: vagga.yaml (hunk 20 above)
Comment (DrEdi) [reply vagga.yaml]: ok

File: vagga.yaml (hunk 30 above)
Comment (DrEdi) [reply vagga.yaml]: Answered in gitter :)

File: vagga.yaml (hunk 31 above)
Comment (DrEdi) [reply vagga.yaml:25]: Ok.

File: vagga.yaml [hunk 34]
```diff
@@ -21,7 +21,7 @@ containers:
         - libpq
       - !Py3Requirements requirements.txt
     environ:
-          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
```
Comment (webknjaz) [inline vagga.yaml]: Why did you remove this var? It suits its purpose well.

File: vagga.yaml [hunk 35]
```diff
@@ -21,7 +21,7 @@ containers:
...
+      environ-file: /work/blog/.env
```
Comment (webknjaz) [inline vagga.yaml]: the indentation is broken here

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: vagga.yaml (hunk 35 above)
Comment (DrEdi) [reply vagga.yaml]: You mean it must be under 'environ' without tab, right?

File: vagga.yaml (hunk 30 above)
Comment (webknjaz) [reply vagga.yaml]: Please add `-k $PGDATA` just in case

File: vagga.yaml (hunk 28 above)
Comment (webknjaz) [reply vagga.yaml]: Still change `/tmp` to `$PGDATA`

**Review by webknjaz — COMMENTED** [review]
Could you please also specify the minimum version of `vagga` needed for this config to work?

File: vagga.yaml [hunk 36]
```diff
@@ -1,26 +1,87 @@
+minimum-vagga: v0.7.0-1-gc613573
```
Comment (webknjaz) [inline vagga.yaml]: I don't think vagga is able to compare commit hash part of the version, please cut it off.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: vagga.yaml (hunk 36 above)
Comment (DrEdi) [reply vagga.yaml]: Oh, I thought it's required.

File: vagga.yaml [hunk 37]
```diff
@@ -1,26 +1,87 @@
+minimum-vagga: v0.7.0-1
```
Comment (webknjaz) [inline vagga.yaml:1]: `v0.7.0`is enough

**Review by webknjaz — APPROVED** [review]
